dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs

```diff
@@ -0,0 +1,508 @@
+
+            using HttpRequestMessage request = DeleteSchemaRequest.Create(className).Build();
+            (HttpResponseMessage response, string responseContent) = await this.ExecuteHttpRequestAsync(request, cancel).ConfigureAwait(false);
+
+            try
+            {
+                response.EnsureSuccessStatusCode();
+            }
+            catch (HttpRequestException e)
+            {
+                this._log.LogError(e, "Collection deletion failed: {0}, {1}", e.Message, responseContent);
+                throw;
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<string> UpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancel = default)
+    {
+        Verify.NotNullOrEmpty(collectionName, "Collection name is empty");
+
+        return await this.UpsertBatchAsync(collectionName, new[] { record }, cancel).FirstOrDefaultAsync(cancel).ConfigureAwait(false) ?? string.Empty;
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> records,
+        [EnumeratorCancellation] CancellationToken cancel = default)
+    {
+        Verify.NotNullOrEmpty(collectionName, "Collection name is empty");
+
+        this._log.LogTrace("Upserting vectors");
+
+        string className = ToWeaviateFriendlyClassName(collectionName);
+        BatchRequest requestBuilder = BatchRequest.Create(className);
+        foreach (MemoryRecord? record in records)
+        {
+            requestBuilder.Add(record);
+        }
+
+        using HttpRequestMessage request = requestBuilder.Build();
+        (HttpResponseMessage response, string responseContent) = await this.ExecuteHttpRequestAsync(request, cancel).ConfigureAwait(false);
+
+        response.EnsureSuccessStatusCode();
+        BatchResponse[]? result = JsonSerializer.Deserialize<BatchResponse[]>(responseContent, s_jsonSerializerOptions);
+        foreach (BatchResponse batchResponse in result!)
+        {
+            yield return batchResponse.Id!;
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<MemoryRecord?> GetAsync(string collectionName, string key, bool withEmbedding = false, CancellationToken cancel = default)
+    {
+        Verify.NotNullOrEmpty(collectionName, "Collection name is empty");
+        Verify.NotNullOrEmpty(key, "Key is empty");
+
+        using HttpRequestMessage request = new GetObjectRequest
+        {
+            Id = key,
+            Additional = withEmbedding ? new[] { "vector" } : null
+        }.Build();
+
+        (HttpResponseMessage response, string responseContent) = await this.ExecuteHttpRequestAsync(request, cancel).ConfigureAwait(false);
+        try
+        {
+            response.EnsureSuccessStatusCode();
+        }
+        catch (HttpRequestException e)
+        {
+            this._log.LogTrace("Request for vector failed {0}, for key {1}", e.Message, key);
```

Comment:
I would suggest not logging the name of the key as there's a chance it could contain PII

Reply:
Line 254 of same class, the `string key` is used as the `id` for looking up the record from Weaviate. Is it expected for PII to be in a key? Is there any documentation on what the parameters are expected to look like / contain?

Reply:
A couple of follow up questions:
- Is the index ID generated service-side? (i.e. what is the batchResponse.Id being returned at line 242?)
- From what I can tell from https://weaviate.io/developers/weaviate/client-libraries/cli, it can be a uuid. Does it have to be a UUID?

Context: 
The key passed to GetAsync and RemoveAsync is expected to be natural language, and could AI-generated. It is possible for this to contain PII. You can see for QdrantMemoryStore, the index ID also has to be a UUID, we had to create separate methods for lookup depending on how the user wants to look up their data. 

*Note that Pinecone allows the index ID to be a provided string of any format.

Reply:
> Is the index ID generated service-side?

It is provided by the user.

> Does it have to be a UUID?

Yes.

> The key passed to GetAsync and RemoveAsync is expected to be natural language

This is not clear and should be called out in the documentation on the method. I can change the code to accommodate this, but Weaviate expects key to be UUID. If a "natural language" key is provided by the framework the only option I have is to set this as a property on the document and attempt to filter on that value on retrieval. This will not guarantee uniqueness though, which is a concern.

Reply:
I have omitted the key from the log.